File: frontend/lib/api/grades.ts
/**
 * Grades API helpers — M3.21 (Essay review interface).
 *
 * Covers:
 *   GET   /essays/{essayId}/grade                      — fetch current grade
 *   PATCH /grades/{gradeId}/feedback                   — edit summary feedback
 *   PATCH /grades/{gradeId}/criteria/{criterionScoreId} — override score or feedback
 *   POST  /grades/{gradeId}/lock                       — lock grade as final
 *
 * Aligned with backend `app/schemas/grade.py` response shapes.
 * Note: `PatchCriterionRequest` intentionally restricts fields to non-null
 * values to match the current backend validator contract.
 *
 * Security notes:
 * - No student PII is logged; only entity IDs appear here.
 * - All endpoints require a valid JWT access token.
 * - Essay content is never stored client-side (no localStorage/sessionStorage).
 */

import { apiGet, apiPatch, apiPost } from "@/lib/api/client";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/** LLM confidence in a criterion score. */
export type ConfidenceLevel = "high" | "medium" | "low";

/** Grading strictness applied at grade-generation time. */
export type StrictnessLevel = "lenient" | "balanced" | "strict";

/**
 * Per-criterion score within a GradeResponse.
 * Matches backend `CriterionScoreResponse` exactly.
 *
 * `rubric_criterion_id` cross-references the criterion in
 * `assignment.rubric_snapshot.criteria[].id` so callers can look up
 * the criterion name, weight, and score range.
 */
export interface CriterionScoreResponse {
  /** UUID of the CriterionScore row. */
  id: string;
  /** UUID of the rubric criterion — used to look up name/weight in the snapshot. */
  rubric_criterion_id: string;
  ai_score: number;
  /** Teacher override — null when no override has been made. */
  teacher_score: number | null;
  /** final_score = teacher_score ?? ai_score */
  final_score: number;
  ai_justification: string;
  /** AI-generated per-criterion student feedback; null for older grades. */
  ai_feedback: string | null;
  /** Teacher-written criterion feedback; null until the teacher edits it. */
  teacher_feedback: string | null;
  confidence: ConfidenceLevel;
  created_at: string;
}

/**
 * Full grade response returned by GET /essays/{essayId}/grade.
 * Matches backend `GradeResponse` exactly.
 */
export interface GradeResponse {
  id: string;
  essay_version_id: string;
  /** Sum of all criterion final_scores. Recalculated after every override.
   *  Serialized as a JSON string from backend Decimal (e.g. "7.00"). */
  total_score: string;
  /** Sum of all criterion max_scores — fixed at grade creation time.
   *  Serialized as a JSON string from backend Decimal (e.g. "10.00"). */
  max_possible_score: string;
  /** AI-generated overall summary feedback. */
  summary_feedback: string;
  /** Teacher-edited summary feedback; null until the teacher edits it. */
  summary_feedback_edited: string | null;
  strictness: StrictnessLevel;
  ai_model: string;
  prompt_version: string;
  is_locked: boolean;
  locked_at: string | null;
  /** Derived from criterion confidence levels: low if any criterion is low;
   *  medium if any is medium (and none is low); high if all are high.
   *  Null for grades written before M4.1. */
  overall_confidence: ConfidenceLevel | null;
  created_at: string;
  criterion_scores: CriterionScoreResponse[];
}

/**
 * Request body for PATCH /grades/{gradeId}/criteria/{criterionScoreId}.
 * At least one field must be provided (validated by backend).
 *
 * Note: the backend validator 422s when both fields are None; sending an
 * explicit JSON `null` maps to None and triggers the same error. Clearing
 * a criterion override/feedback is not supported by the current backend
 * contract — omit the field to leave it unchanged.
 *
 * Matches backend `PatchCriterionRequest`.
 */
export interface PatchCriterionRequest {
  teacher_score?: number;
  teacher_feedback?: string;
}

/**
 * Request body for PATCH /grades/{gradeId}/feedback.
 * Matches backend `PatchFeedbackRequest`.
 */
export interface PatchFeedbackRequest {
  /** min_length: 1, max_length: 10000 (server-enforced). */
  summary_feedback: string;
}

// ---------------------------------------------------------------------------
// API functions
// ---------------------------------------------------------------------------

/**
 * Fetch the current grade for an essay, including all criterion scores.
 * Calls GET /api/v1/essays/{essayId}/grade.
 */
export async function getGrade(essayId: string): Promise<GradeResponse> {
  return apiGet<GradeResponse>(`/essays/${essayId}/grade`);
}

/**
 * Override a criterion score and/or feedback for a grade.
 * Calls PATCH /api/v1/grades/{gradeId}/criteria/{criterionScoreId}.
 *
 * Returns the full updated GradeResponse (including recalculated total_score).
 */
export async function overrideCriterionScore(
  gradeId: string,
  criterionScoreId: string,
  data: PatchCriterionRequest,
): Promise<GradeResponse> {
  return apiPatch<GradeResponse>(
    `/grades/${gradeId}/criteria/${criterionScoreId}`,
    data,
  );
}

/**
 * Update the teacher-edited summary feedback for a grade.
 * Calls PATCH /api/v1/grades/{gradeId}/feedback.
 */
export async function updateFeedback(
  gradeId: string,
  data: PatchFeedbackRequest,
): Promise<GradeResponse> {
  return apiPatch<GradeResponse>(`/grades/${gradeId}/feedback`, data);
}

/**
 * Lock a grade as final. No further edits are allowed after locking.
 * Calls POST /api/v1/grades/{gradeId}/lock.
 *
 * Idempotent — locking an already-locked grade succeeds without error.
 */
export async function lockGrade(gradeId: string): Promise<GradeResponse> {
  return apiPost<GradeResponse>(`/grades/${gradeId}/lock`, {});
}

File: backend/app/routers/grades.py
"""Grades router — grade read, edit, and lock endpoints.

All endpoints require a valid JWT (``get_current_teacher`` dependency).
No student PII is logged — only entity IDs appear in log output.

Endpoints (``essay_grade_router``, prefix ``/essays``):
  GET   /essays/{essayId}/grade              — fetch grade with all criterion scores

Endpoints (``grades_router``, prefix ``/grades``):
  PATCH /grades/{gradeId}/feedback           — update overall summary feedback
  PATCH /grades/{gradeId}/criteria/{criterionId} — override a criterion score or feedback
  POST  /grades/{gradeId}/lock               — lock grade (no further edits allowed)
"""

from __future__ import annotations

import uuid

from fastapi import APIRouter, Depends
from fastapi.responses import JSONResponse

from app.db.session import AsyncSession, get_db
from app.dependencies import get_current_teacher
from app.models.user import User
from app.schemas.grade import PatchCriterionRequest, PatchFeedbackRequest
from app.services.grade import (
    get_grade_audit_log,
    get_grade_for_essay,
    lock_grade,
    override_criterion,
    update_grade_feedback,
)

#: Router for essay-scoped grade retrieval.
essay_grade_router = APIRouter(prefix="/essays", tags=["grades"])

#: Router for grade-level edit and lock operations.
grades_router = APIRouter(prefix="/grades", tags=["grades"])


# ---------------------------------------------------------------------------
# GET /essays/{essayId}/grade
# ---------------------------------------------------------------------------


@essay_grade_router.get(
    "/{essay_id}/grade",
    summary="Get the current grade for an essay with all criterion scores",
)
async def get_essay_grade_endpoint(
    essay_id: uuid.UUID,
    teacher: User = Depends(get_current_teacher),
    db: AsyncSession = Depends(get_db),
) -> JSONResponse:
    """Return the current grade for an essay including all criterion scores.

    Response body: ``{"data": GradeResponse}``

    Returns 403 if the essay belongs to a different teacher.
    Returns 404 if the essay or its grade does not exist.
    """
    grade_response = await get_grade_for_essay(
        db=db,
        essay_id=essay_id,
        teacher_id=teacher.id,
    )
    return JSONResponse(
        status_code=200,
        content={"data": grade_response.model_dump(mode="json")},
    )


# ---------------------------------------------------------------------------
# PATCH /grades/{gradeId}/feedback
# ---------------------------------------------------------------------------


@grades_router.patch(
    "/{grade_id}/feedback",
    summary="Update the summary feedback for a grade",
)
async def patch_grade_feedback_endpoint(
    grade_id: uuid.UUID,
    body: PatchFeedbackRequest,
    teacher: User = Depends(get_current_teacher),
    db: AsyncSession = Depends(get_db),
) -> JSONResponse:
    """Update the teacher-edited summary feedback text for a grade.

    Writes a ``feedback_edited`` audit log entry with before/after values.

    Response body: ``{"data": GradeResponse}``

    Returns 403 if the grade belongs to a different teacher.
    Returns 404 if the grade does not exist.
    Returns 409 if the grade is locked.
    """
    grade_response = await update_grade_feedback(
        db=db,
        grade_id=grade_id,
        teacher_id=teacher.id,
        summary_feedback=body.summary_feedback,
    )
    return JSONResponse(
        status_code=200,
        content={"data": grade_response.model_dump(mode="json")},
    )


# ---------------------------------------------------------------------------
# PATCH /grades/{gradeId}/criteria/{criterionId}
# ---------------------------------------------------------------------------


@grades_router.patch(
    "/{grade_id}/criteria/{criterion_score_id}",
    summary="Override a criterion score or feedback",
)
async def patch_criterion_score_endpoint(
    grade_id: uuid.UUID,
    criterion_score_id: uuid.UUID,
    body: PatchCriterionRequest,
    teacher: User = Depends(get_current_teacher),
    db: AsyncSession = Depends(get_db),
) -> JSONResponse:
    """Override the teacher score and/or feedback for a specific criterion.

    At least one of ``teacher_score`` or ``teacher_feedback`` must be provided.
    Writes a ``score_override`` audit log entry with before/after values.

    Response body: ``{"data": GradeResponse}``

    Returns 403 if the grade belongs to a different teacher.
    Returns 404 if the grade or criterion score does not exist.
    Returns 409 if the grade is locked.
    Returns 422 if neither field is provided.
    """
    grade_response = await override_criterion(
        db=db,
        grade_id=grade_id,
        criterion_score_id=criterion_score_id,
        teacher_id=teacher.id,
        teacher_score=body.teacher_score,
        teacher_feedback=body.teacher_feedback,
    )
    return JSONResponse(
        status_code=200,
        content={"data": grade_response.model_dump(mode="json")},
    )


# ---------------------------------------------------------------------------
# POST /grades/{gradeId}/lock
# ---------------------------------------------------------------------------


@grades_router.post(
    "/{grade_id}/lock",
    status_code=200,
    summary="Lock a grade so no further edits are allowed",
)
async def lock_grade_endpoint(
    grade_id: uuid.UUID,
    teacher: User = Depends(get_current_teacher),
    db: AsyncSession = Depends(get_db),
) -> JSONResponse:
    """Lock a grade as final.

    Sets ``is_locked=True`` and records a ``grade_locked`` audit log entry.
    Locking an already-locked grade is idempotent and succeeds without error.

    Response body: ``{"data": GradeResponse}``

    Returns 403 if the grade belongs to a different teacher.
    Returns 404 if the grade does not exist.
    """
    grade_response = await lock_grade(
        db=db,
        grade_id=grade_id,
        teacher_id=teacher.id,
    )
    return JSONResponse(
        status_code=200,
        content={"data": grade_response.model_dump(mode="json")},
    )


# ---------------------------------------------------------------------------
# GET /grades/{gradeId}/audit
# ---------------------------------------------------------------------------


@grades_router.get(
    "/{grade_id}/audit",
    summary="Get the audit log for a grade",
)
async def get_grade_audit_endpoint(
    grade_id: uuid.UUID,
    teacher: User = Depends(get_current_teacher),
    db: AsyncSession = Depends(get_db),
) -> JSONResponse:
    """Return the full change history for a grade in chronological order.

    Includes all audit entries for the grade itself and its criterion scores
    (e.g. ``feedback_edited``, ``score_override``, ``grade_locked``,
    ``score_clamped``).  Each entry includes timestamp, actor (teacher_id),
    entity type/id, action, and before/after values.

    No student PII is included — only entity IDs appear in the response.

    Response body: ``{"data": [AuditLogEntryResponse, ...]}``

    Returns 403 if the grade belongs to a different teacher.
    Returns 404 if the grade does not exist.
    """
    entries = await get_grade_audit_log(
        db=db,
        grade_id=grade_id,
        teacher_id=teacher.id,
    )
    return JSONResponse(
        status_code=200,
        content={"data": [e.model_dump(mode="json") for e in entries]},
    )
